File: rust/39.combination-sum.rs
/*
 * @lc app=leetcode id=39 lang=rust
 *
 * [39] Combination Sum
 */

// @lc code=start
impl Solution {
    pub fn combination_sum(candidates: Vec<i32>, target: i32) -> Vec<Vec<i32>> {
        let mut combinations = vec![];
        fn backtrack(
            candidates: &Vec<i32>,
            target: i32,
            s: usize,
            combination: &mut Vec<i32>,
            combinations: &mut Vec<Vec<i32>>,
        ) {
            for i in s..candidates.len() {
                if candidates[i] > target {
                    continue;
                }
                combination.push(candidates[i]);
                if candidates[i] == target {
                    combinations.push(combination.clone());
                } else {
                    backtrack(
                        candidates,
                        target - candidates[i],
                        i,
                        combination,
                        combinations,
                    );
                }
                combination.pop();
            }
        }
        backtrack(&candidates, target, 0, &mut vec![], &mut combinations);
        combinations
    }
}
// @lc code=end

File: javascript/39.combination-sum.js
/*
 * @lc app=leetcode id=39 lang=javascript
 *
 * [39] Combination Sum
 */

// @lc code=start
/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */
var combinationSum = function (candidates, target) {
  const combinations = [];
  function backtrack(i, target, combination) {
    for (; i < candidates.length; i++) {
      if (candidates[i] > target) continue;
      combination.push(candidates[i]);
      if (candidates[i] === target) combinations.push([...combination]);
      else backtrack(i, target - candidates[i], combination);
      combination.pop();
    }
  }
  backtrack(0, target, []);
  return combinations;
};
// @lc code=end
